File: src/deno/oak/routes/cache.ts
import { Router, Context } from "../deps.ts";
import { cacheService } from "../services/cache.ts";

// The TTL is part of the response contract and must match what is written
// to Redis. See contracts/rest/canonical-payloads.md.
const CACHE_TTL_SECONDS = 300;

const router = new Router();

router.get("/cache", async (ctx: Context) => {
  const url = ctx.request.url;
  const key = url.searchParams.get("key");
  if (!key) {
    ctx.response.status = 400;
    ctx.response.headers.set("Content-Type", "application/json");
    ctx.response.body = JSON.stringify({ error: "Key parameter is required" });
    return;
  }
  const cached = await cacheService.get(key);
  if (cached) {
    ctx.response.status = 200;
    ctx.response.headers.set("Content-Type", "application/json");
    ctx.response.body = JSON.stringify({ key, value: cached, cached: true, ttl: CACHE_TTL_SECONDS, timestamp: new Date().toISOString() });
    return;
  }
  const value = `Cached value for ${key} at ${new Date().toISOString()}`;
  await cacheService.set(key, value, CACHE_TTL_SECONDS);
  ctx.response.status = 200;
  ctx.response.headers.set("Content-Type", "application/json");
  ctx.response.body = JSON.stringify({ key, value, cached: false, ttl: CACHE_TTL_SECONDS, timestamp: new Date().toISOString() });
});

export { router as cacheRouter };

File: src/go/chi/handlers/cache.go
package handlers

import (
	"encoding/json"
	"net/http"
	"time"

	"chi/services"
)

// The TTL is part of the response contract; it must match the value the
// cache service writes. See contracts/rest/canonical-payloads.md.
const cacheTTLSeconds = 300

func CacheHandler(cacheService *services.CacheService) http.HandlerFunc {
	return func(w http.ResponseWriter, r *http.Request) {
		w.Header().Set("Content-Type", "application/json")

		key := "test"
		if param := r.URL.Query().Get("key"); param != "" {
			key = param
		}

		value, wasCached, err := cacheService.GetOrSet(key)
		if err != nil {
			http.Error(w, "Cache error", http.StatusInternalServerError)
			return
		}

		json.NewEncoder(w).Encode(map[string]interface{}{
			"key":       key,
			"value":     value,
			"cached":    wasCached,
			"ttl":       cacheTTLSeconds,
			"timestamp": time.Now().UTC().Format(time.RFC3339),
		})
	}
}
